src/DataProcessing.py

```diff
@@ -0,0 +1,274 @@
+
+        # Backbone normals
+        n_2 = _normalize(torch.cross(u_2, u_1), dim=-1)
+        n_1 = _normalize(torch.cross(u_1, u_0), dim=-1)
+
+        # Angle between normals
+        cosD = torch.sum(n_2 * n_1, -1)
+        cosD = torch.clamp(cosD, -1 + eps, 1 - eps)
+        D = torch.sign(torch.sum(u_2 * n_1, -1)) * torch.acos(cosD)
+
+        # This scheme will remove phi[0], psi[-1], omega[-1]
+        D = F.pad(D, [1, 2])
+        D = torch.reshape(D, [-1, 3])
+        # Lift angle representations to the circle
+        D_features = torch.cat([torch.cos(D), torch.sin(D)], 1)
+        return D_features
+
+    def _positional_embeddings(self, edge_index: torch.Tensor) -> torch.Tensor:
+        # From https://github.com/jingraham/neurips19-graph-protein-design
+        d = edge_index[0] - edge_index[1]
+
+        frequency = torch.exp(
+            torch.arange(0, self.n_pos_embeddings, 2, dtype=self.dtype, device=self.device)
+            * -(np.log(10000.0) / self.n_pos_embeddings)
+        )
+        angles = d.unsqueeze(-1) * frequency
+        E = torch.cat((torch.cos(angles), torch.sin(angles)), -1)
+        return E
+
+    def _orientations(self, X: torch.Tensor) -> torch.Tensor:
+        forward = _normalize(X[1:] - X[:-1])
+        backward = _normalize(X[:-1] - X[1:])
+        forward = F.pad(forward, [0, 0, 0, 1])
+        backward = F.pad(backward, [0, 0, 1, 0])
+        return torch.cat([forward.unsqueeze(-2), backward.unsqueeze(-2)], -2)
+
+    def _sidechains(self, df: pd.DataFrame) -> torch.Tensor:
+        # For each structure, X should be a num_residues x 4 x 3 nested
+        # list of the positions of the backbone N, C-alpha, C, and O atoms of
+        # each residue (in that order).
+        X = torch.as_tensor(
+            df[df["name"].isin(["N", "CA", "C", "O"])][["x", "y", "z"]].values.reshape(-1, 4, 3)
+        )
+        n, origin, c = X[:, 0], X[:, 1], X[:, 2]
+        c, n = _normalize(c - origin), _normalize(n - origin)
+        bisector = _normalize(c + n)
+        perp = _normalize(torch.cross(c, n))
+        vec = -bisector * math.sqrt(1 / 3) - perp * math.sqrt(2 / 3)
+        return vec
+    
+    def _flexibility(self, protein: struc.AtomArray, cutoff: float, num_modes: int):
+        # Filter CA atoms only
+        ca = protein[(protein.atom_name == "CA") & (protein.element == "C")]
+    
+        # Define force field and GNM object
+        ff = springcraft.InvariantForceField(cutoff_distance= cutoff)
+        gnm = springcraft.GNM(ca, ff)
+
+        if self.type_flexibility == "pseudo":
+            eigenval, eigenvec = gnm.eigen()
+            pseudo_fluc = pseudo_fluctuation_measure(eigenval, eigenvec, num_modes, lambda x: 1/x)
+            pseudo_fluc = torch.reshape(pseudo_fluc, [-1, 1])
+            return pseudo_fluc
+        elif self.type_flexibility == "msqf":
+            msqf = torch.from_numpy(gnm.mean_square_fluctuation(mode_subset = np.array([i for i in range(1, num_modes+1)])))
```

Comment:
Instead of `np.array([i for i in range(1, num_modes+1)])` use `np.arange(1, num_modes+1)` -> more readable